core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala

```diff
@@ -346,19 +348,34 @@ trait AttributeComponent {
         AlterAttributesUsingScratchTableQueries.updateAction(insertFunction(attrsToUpdateMap.values.toSeq))
     }
 
+    //noinspection SqlDialectInspection
     object AlterAttributesUsingScratchTableQueries extends RawSqlQuery {
       val driver: JdbcProfile = AttributeComponent.this.driver
 
       //MySQL seems to handle null safe operators inefficiently. the solution to this
-      //is to use ifnull and default to -2 if the list_index is null. list_index of -2
-      //is never used otherwise
+      //is to use ifnull and default to 0 if the list_index is null.
+
+      // Converting null to the 0 list_index enables converting single attributes to lists, and lists back to single
+      // attributes.
 
       // updateInMasterAction: updates any row in *_ATTRIBUTE that also exists in *_ATTRIBUTE_SCRATCH
-      def updateInMasterAction(transactionId: String) =
-        sql"""update #${baseTableRow.tableName} a
-                join #${baseTableRow.tableName}_SCRATCH ta
-                on (a.namespace,a.name,a.owner_id,ifnull(a.list_index,-2))=(ta.namespace,ta.name,ta.owner_id,ifnull(ta.list_index,-2)) and ta.transaction_id = $transactionId
-                set a.value_string=ta.value_string, a.value_number=ta.value_number, a.value_boolean=ta.value_boolean, a.value_json=ta.value_json, a.value_entity_ref=ta.value_entity_ref, a.list_length=ta.list_length, a.deleted=ta.deleted """.as[Int]
+      def updateInMasterAction(transactionId: String) = {
+          sql"""
+          update #${baseTableRow.tableName} a
+              join #${baseTableRow.tableName}_SCRATCH ta
+              on (a.namespace, a.name, a.owner_id, ifnull(a.list_index, 0)) =
+                 (ta.namespace, ta.name, ta.owner_id, ifnull(ta.list_index, 0))
+                  and ta.transaction_id = $transactionId
+          set a.value_string=ta.value_string,
+              a.value_number=ta.value_number,
+              a.value_boolean=ta.value_boolean,
+              a.value_json=ta.value_json,
+              a.value_entity_ref=ta.value_entity_ref,
+              a.list_index=ta.list_index,
```

Comment:
There doesn't seem to be need to set the `a.list_index` as it is part of join condition just like `namespace` and `name`

Reply:
Hmm... I believe `a.list_index == ta.list_index` makes sure you only update the same rows. Otherwise wouldn't you end up with a cross join, with `a` index 1 matching multiple `ta` rows 1,2,3?

Reply:
Oh I meant that since we already have `a.list_index == ta.list_index` in the join statement, we don't need to update `a.list_index`'s value in the set statement as `a.list_index` and `ta.list_index` will have same value and it would be redundant to update it.

Reply:
Oh, whoops, I thought I was looking at a different part of the query.

No, we don't have `a.list_index == ta.list_index` in the join statement.

We have `ifnull(a.list_index, 0) == ifnull(ta.list_index, 0)`.

The difference is that for a single, aka non-list value, it will have `list_index = null` before the query runs. Afterwards we want that `list_index` to be updated to our new non-null value. This also goes when switching from a list back to a single value.

Reply:
Ah that makes sense. Thank you for the explanation 👍